engine/classes/Elgg/Database/PrivateSettingsTable.php

```diff
@@ -132,29 +134,40 @@ public function getAll($entity_guid) {
 	 * @param int    $entity_guid The entity GUID
 	 * @param string $name        The name of the setting
 	 * @param string $value       The value of the setting
+	 *
 	 * @return bool
+	 * @throws DatabaseException
 	 */
 	public function set($entity_guid, $name, $value) {
-		$this->cache->clear($entity_guid);
-		_elgg_services()->boot->invalidateCache();
-
 		if (!$this->entities->exists($entity_guid)) {
 			return false;
 		}
 
-		$query = "
-			INSERT into {$this->table}
-			(entity_guid, name, value) VALUES
-			(:entity_guid, :name, :value)
-			ON DUPLICATE KEY UPDATE value = :value
-		";
-		$params = [
-			':entity_guid' => (int) $entity_guid,
-			':name' => (string) $name,
-			':value' => (string) $value,
-		];
+		_elgg_invalidate_cache_for_entity($entity_guid);
+
+		$qb = Select::fromTable('private_settings');
+		$qb->select('id')
+			->where($qb->compare('name', '=', $name, ELGG_VALUE_STRING))
+			->andWhere($qb->compare('entity_guid', '=', $entity_guid, ELGG_VALUE_INTEGER));
 
-		$result = $this->db->insertData($query, $params);
+		$row = $this->db->getDataRow($qb);
```

Comment:
Not really efficient. Better strategy would be to cache settings with as database rows and grab IDs from there